src/base/bittorrent/sessionimpl.cpp

```diff
@@ -5033,38 +5032,39 @@ void SessionImpl::loadCategories()
 {
     m_categories.clear();
 
-    QFile confFile {(specialFolderLocation(SpecialFolder::Config) / CATEGORIES_FILE_NAME).data()};
-    if (!confFile.exists())
-    {
-        // TODO: Remove the following upgrade code in v4.5
-        // == BEGIN UPGRADE CODE ==
-        upgradeCategories();
-        m_needUpgradeDownloadPath = true;
-        // == END UPGRADE CODE ==
-
-//        return;
-    }
+    const int fileMaxSize = 1024 * 1024;
+    const Path path = specialFolderLocation(SpecialFolder::Config) / CATEGORIES_FILE_NAME;
 
-    if (!confFile.open(QFile::ReadOnly))
+    const auto readResult = Utils::IO::readFile(path, fileMaxSize);
+    if (!readResult)
     {
-        LogMsg(tr("Failed to load Categories. File: \"%1\". Error: \"%2\"")
-               .arg(confFile.fileName(), confFile.errorString()), Log::CRITICAL);
+        if (readResult.error().status == Utils::IO::ReadError::NotExist)
+        {
+            // TODO: Remove the following upgrade code in v4.5
+            // == BEGIN UPGRADE CODE ==
+            upgradeCategories();
+            m_needUpgradeDownloadPath = true;
+            // == END UPGRADE CODE ==
+//            return;
+        }
+
+        LogMsg(tr("Failed to load Categories. %1").arg(readResult.error().message), Log::WARNING);
```

Comment:
The logic is still broken. In the case of a non-existent "categories" file, it should neither log an error message nor `return`, but just execute the upgrade code above and continue further (re-check how it was done previously).

So this `LogMsg` and `return` below should be under `else` branch.

Reply:
>In the case of a non-existent "categories" file, it should neither log an error message nor return, but just execute the upgrade code above and continue further (re-check how it was done previously).

In previous code, it will get into `if (!confFile.open(QFile::ReadOnly))` isn't?

Reply:
> In previous code, it will get into `if (!confFile.open(QFile::ReadOnly))` isn't?

Yes. But it first creates the file if it wasn't exist previously. So non-existing file wasn't still an error case previously.

Reply:
Done.